performous/performous
# Read any Replay Gain meta-information, use it to normalise the audio "loudness".

NOTE: _I closed my other pull request since it was simply quicker to re-merge the changes manually (with `meld`) rather than muck around in the leavings of `git`.  I had managed to accidentality pull in a huge whack of white-space changes that needed to disappear.  Please accept my apologies._

### What does this PR do?

This change implements loading and use of any [Replay Gain](https://en.wikipedia.org/wiki/ReplayGain) meta information found in the audio files. Files might not have any Replay Gain information in them. To add it, you can use  tools like [loudgain](https://github.com/Moonbase59/loudgain) or [mp3gain](https://mp3gain.sourceforge.net/).  One can easily run `loudgain` over all of Performous' audio tracks using the UNIX "find" command, with the --exec option.

Using the patch generally results in soft audio being made louder and vice-versa.  But don't expect that all music becomes "perfectly" normalised.  A particularly loud piece of music might still be perceived as quite loud - but not enough to warrant manual volume-knob intervention.

It's been tested for many weeks on hundreds of pure-audio and audio-with-video files.

### Closes Issue(s)

Does not close any issues.

### Motivation

When one has audio from different sources, more often than not, the base-level volume of the audio is different. These differences are prevalent even in professionally mastered audio recordings simply due to their jurisdiction. The long-term idea is to somewhat normalise the "loudness" of the audio track, such that continual manual volume adjustments don't need to be made while enjoying Performous.

#### How to Test

- Set your output volume to whatever your normal playback volume is
- Identify some music items in your Performous song list that are either too soft, or too loud
- Play the item in Performous, noting how loud/soft it is
- Use `loudgain` to create the meta-information in the audio file (see below)
- Play the item again in Performous
  - Note the debug log when the gain is found: ffmpeg/debug: readReplayGain() - GAIN IS [-3.73] dB, Linear Gain is [0.65]
  - Listen again to the volume of the track, soft tracks should be made louder, loud tracks softer.
- I found that _generally_ the patch lowers the volume of most tracks sourced from video-viewing websites, since these places normalise the sound to a louder base-level.
- For very formal testing, I guess you could manually adjust the audio volume in Audactiy, and verify it's corrected.

There is a new configuration item in [Audio] that turns the feature on/off.

### More

The libav function that fetches the gain data `av_stream_get_side_data()` is marked as deprecated.  Probably it will need to be migrated to the C++(?) non-deprecated version when all the rest of the API is done.


### Additional Notes

To quickly add Replay Gain meta information, an efficient way is to install "loudgain" (see above), then execute it for every `.mp3` in the music archive.  I've never had an issue with it, but maybe make a backup first.  Best kick this off, watch for a few seconds, then go have a long tea break.

```
$ find . -iname \*mp3 -exec loudgain -q -I3 -S -L -a -k -s e {} \;
$ find . -iname \*m4a -exec loudgain -q -a -k -s e {} \;
```

(Note: last time I looked there was a bug with loudgain and m4a files, but also a patch mentioned in the github comments)
